idp-service/src/main/java/io/mosip/idp/services/AuthorizationHelperService.java

```diff
@@ -0,0 +1,228 @@
+            throw new IdPException(SEND_OTP_FAILED);
+        }
+        return sendOtpResult;
+    }
+
+    protected void validateProvidedAuthFactors(IdPTransaction transaction, List<AuthChallenge> challengeList) throws IdPException {
+        List<List<AuthenticationFactor>> resolvedAuthFactors = authenticationContextClassRefUtil.getAuthFactors(
+                transaction.getRequestedClaims().getId_token().get(ACR).getValues());
+        List<String> providedAuthFactors = challengeList.stream()
+                .map(AuthChallenge::getAuthFactorType)
+                .collect(Collectors.toList());
+
+        boolean result = resolvedAuthFactors.stream().anyMatch( acrFactors ->
+                providedAuthFactors.containsAll(acrFactors.stream().map(AuthenticationFactor::getType).collect(Collectors.toList())));
+
+        if(!result) {
+            log.error("Provided auth-factors {} do not match resolved auth-factor {}", providedAuthFactors, resolvedAuthFactors);
+            throw new IdPException(AUTH_FACTOR_MISMATCH);
+        }
+    }
+
+    protected ResponseWrapper<LinkStatusResponse> createLinkStatusResponse(String transactionId, String status){
+        ResponseWrapper responseWrapper = new ResponseWrapper();
+        LinkStatusResponse linkStatusResponse = new LinkStatusResponse();
+        linkStatusResponse.setTransactionId(transactionId);
+        linkStatusResponse.setLinkStatus(status);
+        responseWrapper.setResponseTime(IdentityProviderUtil.getUTCDateTime());
+        responseWrapper.setResponse(linkStatusResponse);
+        return responseWrapper;
+    }
+
+    protected ResponseWrapper<LinkAuthCodeResponse> getLinkAuthStatusResponse(String authCode) {
+        IdPTransaction idPTransaction = cacheUtilService.getConsentedTransaction(IdentityProviderUtil.generateB64EncodedHash(ALGO_MD5, authCode));
+        if(idPTransaction == null || idPTransaction.getCodeHash() == null || idPTransaction.getRedirectUri() == null)
+            throw new IdPException(ErrorConstants.INVALID_STATE);
+
+        return createLinkAuthCodeResponse(idPTransaction, authCode);
+    }
+
+    protected ResponseWrapper<LinkAuthCodeResponse> createLinkAuthCodeResponse(IdPTransaction idPTransaction, String authCode){
+        ResponseWrapper responseWrapper = new ResponseWrapper();
+        LinkAuthCodeResponse linkAuthCodeResponse = new LinkAuthCodeResponse();
+        linkAuthCodeResponse.setNonce(idPTransaction.getNonce());
+        linkAuthCodeResponse.setState(idPTransaction.getState());
+        linkAuthCodeResponse.setRedirectUri(idPTransaction.getRedirectUri());
+        linkAuthCodeResponse.setCode(authCode);
+        responseWrapper.setResponseTime(IdentityProviderUtil.getUTCDateTime());
+        responseWrapper.setResponse(linkAuthCodeResponse);
+        return responseWrapper;
+    }
+
+    protected String getCacheKey(@NotNull String value) {
+        return IdentityProviderUtil.generateB64EncodedHash(ALGO_MD5, value);
```

Comment:
I think md5 has some vulnerabilities, lets consult with Taheer for the apt high performing algorithm for this need.

Reply:
Changed to SHA3-256